File: src/main/kotlin/com/babasama/edendale/android/data/WyzieKeyStore.kt
package com.babasama.edendale.android.data

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.babasama.edendale.WyzieSecrets
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Securely stores the viewer's Wyzie API key.
 *
 * Everything is lazy so construction never performs a Keystore round trip on
 * the main thread. Callers read or mutate the store from a background thread.
 */
class WyzieKeyStore(context: Context) {

    private val appContext = context.applicationContext

    private val masterKey by lazy {
        MasterKey.Builder(appContext)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
    }

    private val sharedPrefs: SharedPreferences by lazy {
        EncryptedSharedPreferences.create(
            appContext,
            "edendale_wyzie",
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
        )
    }

    val buildKey: String
        get() = WyzieSecrets.apiKey.trim()

    fun userKey(): String = sharedPrefs.getString(KEY_API_KEY, null)?.trim().orEmpty()

    fun resolvedKey(): String = userKey().ifEmpty { buildKey }

    fun isConfigured(): Boolean = resolvedKey().isNotEmpty()

    fun hasUserKey(): Boolean = userKey().isNotEmpty()

    fun usesBuildKey(): Boolean = !hasUserKey() && buildKey.isNotEmpty()

    suspend fun save(key: String) {
        val trimmed = key.trim()
        if (trimmed.isEmpty()) {
            clear()
            return
        }
        withContext(Dispatchers.IO) {
            sharedPrefs.edit().putString(KEY_API_KEY, trimmed).commit()
        }
    }

    suspend fun clear() {
        withContext(Dispatchers.IO) {
            sharedPrefs.edit().remove(KEY_API_KEY).commit()
        }
    }

    private companion object {
        const val KEY_API_KEY = "api_key"
    }
}

File: Shared/Controllers/WyzieKeyStore.swift
//
//  WyzieKeyStore.swift
//  Edendale
//
//  Resolves a user-provided Wyzie API key ahead of the optional build-time
//  key. The user key is stored in the synchronized keychain and is never
//  logged.
//

import Foundation

/// Resolves the optional Wyzie credential. A user-provided key is stored in
/// the synchronized keychain and is never logged.
@MainActor
@Observable
final class WyzieKeyStore {
    private(set) var userKey: String

    init() {
        userKey = KeychainStore.shared.data(for: .wyzieAPIKey)
            .flatMap { String(data: $0, encoding: .utf8) }?
            .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    }

    var buildKey: String {
        ((Bundle.main.object(forInfoDictionaryKey: "WyzieAPIKey") as? String) ?? "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    var resolvedKey: String {
        userKey.isEmpty ? buildKey : userKey
    }

    var isConfigured: Bool { !resolvedKey.isEmpty }
    var hasUserKey: Bool { !userKey.isEmpty }
    var usesBuildKey: Bool { userKey.isEmpty && !buildKey.isEmpty }

    func save(_ key: String) throws {
        let trimmed = key.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            clear()
            return
        }
        try KeychainStore.shared.set(Data(trimmed.utf8), for: .wyzieAPIKey)
        userKey = trimmed
    }

    func clear() {
        KeychainStore.shared.remove(.wyzieAPIKey)
        userKey = ""
    }
}
